### PR TITLE
Linux: Rewrite the event loop using calloop

### PR DESCRIPTION
**Pull request by witelokk**
This PR unifies the event loop code for Wayland and X11. On Wayland, blocking dispatch is now used. On X11, the invisible window is no longer needed.

Release Notes:

- N/A

### CODE REVIEW

File: crates/gpui/src/platform/linux/x11/client.rs [hunk 1]
```diff
@@ -46,192 +59,217 @@ impl X11Client {
...
+                        client.handle_event(event);
+                    }
+                    calloop::channel::Event::Closed => {}
+                })
+                .unwrap();
+        }
+
+        {
+            let xcb_connection = Arc::clone(&xcb_connection);
+            std::thread::spawn(move || loop {
```
Comment (kvark) [inline crates/gpui/src/platform/linux/x11/client.rs:90]: With this change, we have an extra thread to poll events from the windowing system. It also means that each message has to go through one more step: first get from X11 into this thread, then get from this thread over the channel into the `calloop` event loop, and then be processed.

**Review by kvark — COMMENTED** [review]
I'd like to have us on the same page about event loops. Technically speaking, we could have a platform-specific event loop (that's the current code in `main`), or we could have a platform-abstract event loop (like in `winit`). Having something in-between - an abstract-ish event loop that's only used for Linux - seems like a strange design choice.
Is the intent here to also use `calloop` on macOS and Windows eventually?

**Review by kvark — DISMISSED** [review]
a few more concerns

**Comment by mikayla-maki** [comment]
> Having something in-between - an abstract-ish event loop that's only used for Linux - seems like a strange design choice.

I think having a reasonably thin eventloop adapter that meshes with Zed's callback style and provides a bit of structure to other code we write could be good. And it would be nice to not have the invisible window. But that said, I haven't looked too deeply yet :)

File: crates/gpui/src/platform/linux/x11/client.rs (hunk 1 above)
Comment (tadeokondrak) [reply crates/gpui/src/platform/linux/x11/client.rs:90]: See the comments in https://github.com/Smithay/smithay/blob/a04d9ac8441e19e87a6af6410d0f1ddbf4b293c5/src/utils/x11rb.rs for some reasoning for this choice

**Comment by kvark** [comment]
Discussed this [on discord](https://discordapp.com/channels/869392257814519848/1204679850208657418/1210840532381728808) and got a better understanding about how `calloop` fits in. I think it's a good direction, thanks @tadeokondrak for explanation.
Please rebase the PR and address the review points. I think X11 may start working then.

File: crates/gpui/src/platform/linux/x11/window.rs [hunk 2]
```diff
@@ -268,31 +271,34 @@ impl X11WindowState {
...
             fun();
         }
         self.xcb_connection.flush().unwrap();
     }
 
     pub fn refresh(&self) {
-        let mut cb = self.callbacks.lock();
-        if let Some(ref mut fun) = cb.request_frame {
+        let mut cb = self.callbacks.borrow_mut();
+        if let Some(mut fun) = cb.request_frame.take() {
```
Comment (kvark) [inline crates/gpui/src/platform/linux/x11/window.rs:289]: what was the motivation for doing this? did we end up calling a callback within a callback?

**Pull request merged**